src/nmmodel.cpp

```diff
@@ -1227,8 +1227,6 @@ QVariant NmModel::dataRole<NmModel::IconTypeRole>(const QModelIndex & index) con
                 {
                     case NetworkManager::ConnectionSettings::Wireless:
                     case NetworkManager::ConnectionSettings::Wimax:
-                        if (0 > state)
```

Comment:
No. `state` is an integer... and can be 0.

Reply:
State is NetworkManager::ActiveConnection::State
That is enum State { Unknown = 0, Activating, Activated, Deactivating,   Deactivated } according to [documentation](https://api.kde.org/frameworks/networkmanager-qt/html/classNetworkManager_1_1ActiveConnection.html)
It can be zero, but can NEVER be less then zero
if (0 > state) could be true only on negative state that is impossible
Thus, the condition is always false

Reply:
Moreover, according to c++ standard
> `The underlying type of an enumeration is an integral type that can represent all the enumerator values defined in the enumeration. It is implementation-defined which integral type is used as the underlying type for an enumeration except that the underlying type shall not be larger than int unless the value of an enumerator cannot fit in an int or unsigned int.`
Enumerator can be both signed or unsigned on different systems. This means that, for example, "Activating" can be either signed or unsigned. Given that, this code is undefined behavior and it can cause bugs on some systems

Reply:
OK. But then the "problem" is the cast (and auto deduction) done here:
```c++
                  auto state = static_cast<NetworkManager::ActiveConnection::State>(dataRole<ActiveConnectionStateRole>(index).toInt());
```
as the the `dataRole()` returns -1 for "invalid" data types:

```c++
  template <>
  QVariant NmModel::dataRole<NmModel::ActiveConnectionStateRole>(const QModelIndex & index) const
  {
      switch (static_cast<ItemId>(index.internalId()))
      {
          case ITEM_ROOT:
          case ITEM_ACTIVE:
          case ITEM_CONNECTION:
          case ITEM_DEVICE:
          case ITEM_CONNECTION_LEAF:
          case ITEM_DEVICE_LEAF:
          case ITEM_WIFINET:
          case ITEM_WIFINET_LEAF:
              return -1;
          case ITEM_ACTIVE_LEAF:
              return d->mActiveConns[index.row()]->state();
      }
      return QVariant{};
  }
```

But definitely not the condition for "validness" of `state`.

Reply:
Why in hell can it ever be -1? It is enum so you should not compare it against integer values.
Moreover, in gcc/g++ enum is unsigned so the condition is always false.

Reply:
> as the the dataRole() returns -1 for "invalid" data types
That is undefined behavior then casted to NetworkManager::ActiveConnection::State
Anyway `if (0 > state)` is always false on most systems so something should be changed here.

Reply:
I did change the logic for handling `QVariant`s ... please rebase.

Reply:
I am not sure I know how to rebase it(
Every time I try it I get a mess